internal/innerConcat.js

```diff
@@ -0,0 +1,26 @@
+/** @license ISC License (c) copyright 2017 original and current authors */
+/** @author Ian Hofmann-Hicks (evil) */
+
+const curry = require('../helpers/curry')
+const isSameType = require('../predicates/isSameType')
+const isSemigroup = require('../predicates/isSemigroup')
+
+function innerConcat(type, m) {
+  const t = type.type()
+
+  return function(left) {
+    if(!isSemigroup(left)) {
+      throw new TypeError(`${t}.concat: Both containers must contain Semigroups of the same type`)
+    }
+
+    return m.map(right => {
+      if(!isSameType(left, right)) {
+        throw new TypeError(`${t}.concat: Both containers must contain Semigroups of the same type`)
+      }
```

Comment:
`M(String).concat(M(number))` or `M(Array).concat(M(number))` should work, so should be:
```
const rightT = left.concat(right)
    return m.map(right => {
       if(!isSameType(left, rightT)) {
         throw new TypeError(`${t}.concat: Both containers must contain Semigroups of the same type`)
       }
```
What is your thought?